File: annotation/src/main/java/com/tbb/vip/annotation/MainActivity.java
package com.tbb.vip.annotation;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class MainActivity extends AppCompatActivity {

    String TAG = "MainActivity";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        User user = new User();
        Class<User> userClass = User.class;
        try {
            Method method = userClass.getMethod("getUser", String.class, String.class);
            method.invoke(user,"李四","33");


            if (method.isAnnotationPresent(MyAnnotation.class)){
                MyAnnotation annotation = method.getAnnotation(MyAnnotation.class);
                String[] value = annotation.value();
                for (String v:
                     value) {
                    Log.e(TAG,"v->" + v);
                }
            }else {
                    Log.e(TAG,"v->没有这个注解");

            }


            method.invoke(user.aaa,user.age,user.name,"李四","名字","年龄");



        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }

    }

    class User{
        @MyAnnotation("sss")
        String aaa;
        String name;
        String age;


        @MyAnnotation(value = {"张三","23"})
        public void getUser(String name,String age){
            Log.e(TAG,"name" + name + "|age = " + age);
        }
    }


}

File: mvvm/src/main/java/com/wbj/mvvm/MainActivity.kt
package com.wbj.mvvm

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.util.Log
import androidx.databinding.DataBindingUtil
import com.wbj.mvvm.databinding.MainActivityBinding
import com.wbj.mvvm.model.User

class MainActivity : AppCompatActivity() {

//    var binding : MainActivityBinding ?= null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = DataBindingUtil.setContentView<MainActivityBinding>(this,R.layout.activity_main)
        var user = User("涨","停")
        binding.user = User("涨","停")
        binding.isShow = false
        val name = with(user,{
            this.firstName = "jfjfjfj"
            this.lastName = "jfjfjfj"
        })
        Log.e("MainActivity","first-")
        Handler().postDelayed({
//            val user = binding.user//更新值无用
//            user?.firstName = "吾问无为谓"
//            user?.lastName = "少时诵诗书"
            binding.user = User("吾问无为谓","少时诵诗书")
        },2000)


    }
}